### ISSUE TITLE
Move Envoy out of the repository

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#5212]
Having the Envoy changes in the code base has proven to be difficult to manage:
* Base images need to be constantly rebased and pushed into various CI base images
* Local compilation of Cilium is time consuming
* The vagrant box developer image is huge because of the required cache

There are relatively few Envoy side changes at this point. The majority of the work is in the golang side of how the Envoy filters are being configured.

*Proposal:*

1. Move the `envoy/` directory into a separate `cilium/envoy-cilium` repository
2. Set up an automated build on quay.io to automatically build tags.
3. Tag releases after changes as needed
4. Refer to the tag in the Cilium codebase
5. When building the cilium container image, pull in the `cilium/envoy-cilium` container layer and copy over the binary.
6. Continue running the CI as-is but no longer require to compile Envoy for every Cilium build.

**Comment by rlenglet** [comment]
> There are relatively few Envoy side changes at this point. The majority of the work is in the golang side of how the Envoy filters are being configured.

It is actually the opposite. The Cilium Agent / xDS side has been very stable. But we've had to update the Envoy version frequently due to Istio updates, Envoy bug fixes, etc. And that is the cause of the problem.

I'm all for moving Envoy into its own repo and building Envoy into a separate Docker image/layer.

Istio has a similar repo structure, except that they store & download Envoy binaries as-is, not as Docker images. I prefer to deliver them as Docker images, because it will allow us to ship all the Envoy deliverables as one consistent Docker image, which would contain all of:
- the Bazel cache
- all the various builds of the Envoy binaries, for test, production, and istio

Note that shipping an up-to-date Bazel cache is still necessary because it is required to generate the Go protobuf files for the Cilium agent.

**Comment by rlenglet** [comment]
We'll also need to decide in which repo the future Go parsers will live. Those will have dependencies on both Envoy and Cilium.

**Comment by rlenglet** [comment]
@jrajahalme ^^^

**Comment by tgraf** [comment]
> It is actually the opposite. The Cilium Agent / xDS side has been very stable. But we've had to update the Envoy version frequently due to Istio updates, Envoy bug fixes, etc. And that is the cause of the problem.

OK. Let me rephrase, we are not actively changing the agent/xds side in parallel to the Cilium Envoy filter in every other commit these days. It's rebases and targeted bugfixes. We have moved to a relationship where a Cilium version can depend on a particular version of cilium-envoy.



> Istio has a similar repo structure, except that they store & download Envoy binaries as-is, not as Docker images. I prefer to deliver them as Docker images, because it will allow us to ship all the Envoy deliverables as one consistent Docker image

👍 

> Note that shipping an up-to-date Bazel cache is still necessary because it is required to generate the Go protobuf files for the Cilium agent.

Yes. The open question I have is how we develop Envoy in the future. Do we want a flag to the `start.sh` script which uses a different base image that includes the bazel cache? This would allow to continue using the vagrant box as-is but would prevent from every developer having to care about the bazel cache.

> We'll also need to decide in which repo the future Go parsers will live. Those will have dependencies on both Envoy and Cilium.

Ideally they can live in separate repos. They will depend on certain Envoy versions I assume. Will need input from @jrajahalme how extendable and backwards compatible the current interface is and how that will stabilize.

**Comment by rlenglet** [comment]
I have bigger concerns about the Vagrant VM dev workflow. We'll have to pull the artifacts out of the new Docker image in order to make development and dev testing possible in Vagrant.

A minor action item is also to remove the Envoy version hash check from Cilium.

**Comment by jrajahalme** [comment]
The current (Proof-of-Concept) Envoy/Go filter interface is pretty minimal, and if we keep it that way it should be possible to keep it very stable. If we start exposing more of the Envoy feature set to the Go filters then this would change.

Also, the current working model I have is to build the Go filters as a shared library. This makes it possible to keep them in the Cilium repo even if the Envoy image is moved elsewhere. I.e., they can be built separately and paired with any compatible Envoy image via filter configuration. We probably need have an explicit version check to the initial handshake between Envoy and the shared library, though.

**Comment by vadorovsky** [comment]
Big +1. Having Envoy in the same repository and build system makes RPM packaging harder. Possibility of building `cilium-envoy` binary separately (because, if I understand correctly, that would be one of goals?) and building filters as a shared lib would make it much easier.

**Comment by tgraf** [comment]
We should land this early on in the 1.4 development cycle.

**Comment by tgraf** [comment]
This has been completed for 1.4.
